### ISSUE TITLE
Be more lenient on tag parsing

### ISSUE THREAD

**Issue by Adam5Wu** [Zoltu/Gradle.Plugin.Versioning#1]
Great work, very easy to use and good cross-platform compatibility! :D

Could you please be a little more lenient on tag parsing?
I am trying to compile a fork of Google ProtocolBuffer, and I try to use their tagging style, but this plug-in fails when parsing version tags such as:

```
$ git describe
v3.0.0-beta-2-88-gb13874d
```

I see in your code you are trying to match:
v(numbers1).(numbers2)-(numbers3)-g(hash)

How about relax a little bit and match the following:
v(token1).(token2)-(numbers)-g(hash)
In case of the above version tag:
token1 = 3
token2 = 0.0-beta-2
numbers = 88
hash = b13874d

And the version generated will be:
3.0.0-beta-2.88

This pattern should do the trick:

```
v([^.]+)\.(.*)\-([0-9]+)\-g(.*)
```

**Comment by MicahZoltu** [comment]
If I understand you correctly, you are proposing that the minor version would be everything after the major version and before the patch version (commit count)?  That _feels_ a little wrong to me since I'm not a fan of non-numbers as part of the version.  In the example you provided it seems like you really have `v{major}.{minor}.{patch}-{tag name}.{build}-g{hash}`.

How would you feel if the tool spit out `3.0.0.88` for the version given your example?  Do you find the `beta-2` bit to be valuable?  What about `3.0.0.88-beta-2`?

**Comment by MicahZoltu** [comment]
Alternatively (and more opinionated), what if it spit out `3.0.88`, effectively ignoring all of the other stuff in the tag and just using the major/minor?  There is a concern that people using tags like this could end up with duplicate version numbers since these two would evaluate to the same thing: `3.0.0-beta-2-88-g1234abcd` and `3.0.0-beta3-88-g5678defg` (`3.0.88` in both cases).

**Comment by Adam5Wu** [comment]
I am all pro for simplicity. But pure numbers are sometimes too limited in terms of expressiveness.

It seems Google adopts the "semantic versioning" schema, as described here: http://semver.org/

It is in format major.minor.patch, so for the tag
v3.0.0-beta-2
Major = v3
Minor = 0
Patch = 0-beta-2

It is not a perfect schema, but it solves most practical versioning needs.

By relaxing the parser, it not only didn't disable the use of pure numbers for versions, but also accommodates other versioning styles. Seems to me an all-pro-no-con move. :D

**Comment by MicahZoltu** [comment]
I agree, this plugin should support semantic versioning.  However, according to the spec (http://semver.org/#spec-item-9) the tag `v3.2.4-beta-1.5` is as follows:
- Major: `3`
- Minor: `2`
- Patch: `1`
- Tag 1: `beta-1`
- Tag 2: `5`

When adding support for semver, I suppose the right thing to do is support it as close to spec as possible.  This means that the commit count should be used as a trailing tag, after all other tags.  In the case of a git tag like `3.2.4.88-beta-1.5`, followed by 88 commits, this would result in `3.2.4-beta-1.5.88`, where `88` is the third (last) tag.  This comes from the template string `{major}.{minor}.{patch}-{tag1}.{tag2}.{tag3}`.

I'll work on adding this support, thanks for the suggestion!

**Comment by MicahZoltu** [comment]
I didn't implement this quite right in that PR... will open another with a fix.

**Comment by MicahZoltu** [comment]
Closed again by #3

**Comment by MicahZoltu** [comment]
The latest version (major version bump to v3.x) removes semantic versioning out-of-the-box in favor of allowing users to fully customize the way version info is extracted.  Out-of-the-box now only supports versioning like `v1.2` with optional tags.

If you would like to continue using this plugin and leverage semantic versioning, you simply need to include a custom describe processor.  Details can be found in the readme, or you can see an example `build.gradle` file in the tests (the same tests passed before without a custom describe processor, so the regex in these tests should match the previous semantic versioning scheme, allowing you to easily copy/paste the block into your `build.gradle` without any changes.
https://github.com/Zoltu/Gradle.Plugin.Versioning/blob/master/src/test/kotlin/FunctionalTests.kt#L107-L112
https://github.com/Zoltu/Gradle.Plugin.Versioning/blob/master/src/test/kotlin/FunctionalTests.kt#L87-L92
